src/memray/reporters/flamegraph.py

```diff
@@ -75,6 +112,97 @@ def __init__(
         self.data = data
         self.memory_records = memory_records
 
+    @classmethod
+    def update_import_system_frames_for_inverted_flamegraph(
+        cls, nodes: Sequence[FrameNodeDict], node_index: int
+    ) -> bool:
+        if not nodes or not nodes[node_index]:
+            return False
+
+        for child_index in nodes[node_index]["children"]:
+            result = cls.update_import_system_frames_for_inverted_flamegraph(
+                nodes, child_index
+            )
+            nodes[node_index]["import_system"] |= result
+
+        return nodes[node_index]["import_system"]
+
+    @classmethod
+    def generate_nodes(
+        cls,
+        frames: List[FrameNodeDict],
+        all_strings: StringRegistry,
+        interval_list: Sequence[Tuple[int, Union[int, None], int, int, int]],
+    ) -> FlameGraphNodeDict:
+        nodes = cast(FlameGraphNodeDict, collections.defaultdict(list))
+        for frame in frames:
+            nodes["name"].append(all_strings.register(frame["name"]))
+            nodes["function"].append(all_strings.register(frame["location"][0]))
+            nodes["filename"].append(all_strings.register(frame["location"][1]))
+            nodes["lineno"].append(frame["location"][2])
+            nodes["children"].append(frame["children"])
+            if not interval_list:
+                nodes["value"].append(frame["value"])
+                nodes["n_allocations"].append(frame["n_allocations"])
+            nodes["thread_id"].append(all_strings.register(frame["thread_id"]))
+            nodes["interesting"].append(int(frame["interesting"]))
+            nodes["import_system"].append(int(frame["import_system"]))
+        return nodes
+
+    @classmethod
+    def generate_frames(
+        cls,
+        stack_it: List[PythonStackElement],
+        frames: List[FrameNodeDict],
+        node_index_by_key: Dict[Tuple[int, StackFrame, str], int],
+        record: Union[AllocationRecord, TemporalAllocationRecord],
+        current_frame_id: int,
+        inverted: bool,
+        is_inverted_import_system: bool,
+    ) -> List[FrameNodeDict]:
+        current_frame = frames[0]
+        num_skipped_frames = 0
+        is_import_system = False
+
+        for index, stack_frame in stack_it:
+            node_key = (current_frame_id, stack_frame, record.thread_name)
+            if node_key not in node_index_by_key:
+                if is_cpython_internal(stack_frame):
+                    num_skipped_frames += 1
+                    continue
+                # update import_system only if we're generting normal flamegraph
+                # and inverted flamegrap
+                # we don't need to calculate is_frame_from_import_system for inverted
+                # import system tree again
+                if (inverted and not is_inverted_import_system) or not is_import_system:
+                    is_import_system = is_frame_from_import_system(stack_frame)
```

Comment:
I think you should just skip this call whenever `inverted` is true. When `inverted and is_inverted_import_system`, the caller already stripped the import system frames, so we know `is_frame_from_import_system` would always return false. But when `inverted and not is_inverted_import_system` - that is, the full inverted tree without import system nodes dropped - we know that this question isn't answerable. The flag doesn't make sense, because all the call paths that lead to one function are children of the same node. Taking my example from the other day, it doesn't make sense to flag `malloc` as either `is_import_system = True` or as `is_import_system = False`, because half of its callers are import system calls and half aren't. I think you should just skip this check entirely when `inverted is True`, and just leave every node with `is_import_system = False`, because we can't give any other answer accurately.
```suggestion
                if not inverted and not is_import_system:
                    is_import_system = is_frame_from_import_system(stack_frame)
```